File: projects/native-apps/ios/App/LaunchIntroSeen.swift
import Foundation

struct LaunchIntroSeen {
	private let defaults: UserDefaults

	private enum Key {
		static let seen = "launchIntro.seen"
	}

	init(defaults: UserDefaults) {
		self.defaults = defaults
	}

	func claim() -> Bool {
		guard !defaults.bool(forKey: Key.seen) else { return false }
		defaults.set(true, forKey: Key.seen)
		return true
	}
}

File: projects/native-apps/android/app/src/main/kotlin/com/readplace/android/app/LaunchIntroSeen.kt
package com.readplace.android.app

/**
 * The launch intro's persisted flags, standing in for the iOS `UserDefaults`. A
 * read answers null for a key never written, so each flag decides for itself how
 * an absent value reads. Production is SharedPreferences-backed; tests supply
 * their own.
 */
interface KeyValueFlags {
	fun getBoolean(key: String): Boolean?

	fun putBoolean(key: String, value: Boolean)
}

class LaunchIntroSeen(private val flags: KeyValueFlags) {
	fun claim(): Boolean {
		if (flags.getBoolean(SEEN_KEY) == true) return false
		flags.putBoolean(SEEN_KEY, true)
		return true
	}

	private companion object {
		const val SEEN_KEY = "launchIntro.seen"
	}
}
